src/libexpr/primops/fromTOML.cc

```diff
@@ -1,93 +1,87 @@
+        switch(t.type())
+        {
+            case toml::value_t::table:
+                {
+                    auto table = toml::get<toml::table>(t);
 
-                if (auto i2 = i.second->as_table_array()) {
-                    size_t size2 = i2->get().size();
-                    state.mkList(v2, size2);
-                    for (size_t j = 0; j < size2; ++j)
-                        visit(*(v2.listElems()[j] = state.allocValue()), i2->get()[j]);
-                }
-                else
-                    visit(v2, i.second);
-            }
-
-            v.attrs->sort();
-        }
+                    size_t size = 0;
+                    for (auto & i : table) { (void) i; size++; }
 
-        else if (auto t2 = t->as_array()) {
-            size_t size = t2->get().size();
+                    state.mkAttrs(v, size);
 
-            state.mkList(v, size);
+                    for(auto & elem: table) {
 
-            for (size_t i = 0; i < size; ++i)
-                visit(*(v.listElems()[i] = state.allocValue()), t2->get()[i]);
-        }
+                        auto & v2 = *state.allocAttr(v, state.symbols.create(elem.first));
 
-        // Handle cases like 'a = [[{ a = true }]]', which IMHO should be
-        // parsed as a array containing an array containing a table,
-        // but instead are parsed as an array containing a table array
-        // containing a table.
-        else if (auto t2 = t->as_table_array()) {
-            size_t size = t2->get().size();
-
-            state.mkList(v, size);
-
-            for (size_t j = 0; j < size; ++j)
-                visit(*(v.listElems()[j] = state.allocValue()), t2->get()[j]);
-        }
+                        // TODO: note about creating children here in old code
+                        visit(v2, elem.second);
+                    }
+                }
+                break;;
+            case toml::value_t::array:
+                {
+                    auto array = toml::get<std::vector<toml::value>>(t);
+
+                    size_t size = array.size();
+                    state.mkList(v, size);
+                    for (size_t i = 0; i < size; ++i)
+                        visit(*(v.listElems()[i] = state.allocValue()), array[i]);
+                }
+                break;;
+            case toml::value_t::boolean:
+                mkBool(v, toml::get<bool>(t));
+                break;;
+            case toml::value_t::integer:
+                mkInt(v, toml::get<int>(t));
+                break;;
+            case toml::value_t::floating:
+                mkFloat(v, toml::get<double>(t));
+                break;;
+            case toml::value_t::string:
+                mkString(v, toml::get<std::string>(t));
+                break;;
+            case toml::value_t::local_datetime:
+            case toml::value_t::offset_datetime:
+            case toml::value_t::local_date:
+            case toml::value_t::local_time:
+                // TODO: convert to string?
```

Comment:
How should we handle dates and times best? Failure is an option, but we could also convert them to strings.